Makefile

```diff
@@ -1,28 +1,254 @@
+TOP := ./
+TINYUSB_PATH ?= $(TOP)/lib/tinyusb
 
-INC += \
-	hw \
-	hw/bsp \
-	hw/bsp/$(BOARD) \
-	src \
-	lib/tinyusb/tools \
-	_build/build-$(BOARD)
+# Set to 1 to get noisy makefile output
+V ?= 0
 
-SRC_C = \
-	$(addprefix $(CURRENT_PATH)/, $(wildcard src/*.c))
+#-------------- Select the board to build for. ------------
 
-CFLAGS += -Wno-unused-parameter
+BOARD_LIST = $(sort $(subst /.,,$(subst $(TOP)/hw/bsp/,,$(wildcard $(TOP)/hw/bsp/*/.))))
+
+ifeq ($(filter $(BOARD),$(BOARD_LIST)),)
+  $(info You must provide a BOARD parameter with 'BOARD=', supported boards are:)
+  $(foreach b,$(BOARD_LIST),$(info - $(b)))
+  $(error Invalid BOARD specified)
+endif
+
+# Handy check parameter function
+check_defined = \
+    $(strip $(foreach 1,$1, \
+    $(call __check_defined,$1,$(strip $(value 2)))))
+__check_defined = \
+    $(if $(value $1),, \
+    $(error Undefined make flag: $1$(if $2, ($2))))
+
+# Build directory
+BUILD = _build/$(BOARD)
+
+# Get Board identifier and specific defines
+include $(TOP)/hw/bsp/$(BOARD)/board.mk
+
+#-------------- Source files generic across all targets --------------
+
+SRC_C += $(subst $(TOP)/,,$(wildcard $(TOP)/hw/bsp/$(BOARD)/*.c))
+SRC_C += $(subst $(TOP)/,,$(wildcard $(TOP)/src/*.c))
+SRC_C += $(subst $(TOP)/,,$(wildcard $(TOP)/hw/chip/$(TUF2_CHIP_FAMILY)/*.c))
+
+# TinyUSB Stack source
+SRC_C += \
+	$(TINYUSB_PATH)/src/tusb.c \
+	$(TINYUSB_PATH)/src/common/tusb_fifo.c \
+	$(TINYUSB_PATH)/src/device/usbd.c \
+	$(TINYUSB_PATH)/src/device/usbd_control.c \
+	$(TINYUSB_PATH)/src/class/cdc/cdc_device.c \
+	$(TINYUSB_PATH)/src/class/dfu/dfu_rt_device.c \
+	$(TINYUSB_PATH)/src/class/hid/hid_device.c \
+	$(TINYUSB_PATH)/src/class/midi/midi_device.c \
+	$(TINYUSB_PATH)/src/class/msc/msc_device.c \
+	$(TINYUSB_PATH)/src/class/net/net_device.c \
+	$(TINYUSB_PATH)/src/class/usbtmc/usbtmc_device.c \
+	$(TINYUSB_PATH)/src/class/vendor/vendor_device.c \
+	$(TINYUSB_PATH)/src/portable/$(VENDOR)/$(CHIP_FAMILY)/dcd_$(CHIP_FAMILY).c
+
+# Compiler Flags which are generic across all targets
+CFLAGS += \
+        -fdata-sections \
+        -ffunction-sections \
+        -fsingle-precision-constant \
+        -fno-strict-aliasing \
+        -Wdouble-promotion \
+        -Wstrict-prototypes \
+        -Wall \
+        -Wextra \
+        -Werror \
+        -Werror-implicit-function-declaration \
+        -Wfloat-equal \
+        -Wundef \
+        -Wshadow \
+        -Wwrite-strings \
+        -Wsign-compare \
+        -Wmissing-format-attribute \
+	-Wno-unused-parameter \
+        -Wunreachable-code \
+	-ffreestanding
+
+# Compiler flags for defining UF2 configuration
+CFLAGS += \
+	-DAPP_START_ADDRESS=$(APP_START_ADDRESS) \
+	-DBOARD_FLASH_BASE=$(BOARD_FLASH_BASE) \
+	-DUF2_FAMILY=$(UF2_FAMILY) \
+	-DCFG_TUSB_MCU=$(TUSB_MCU)
+
+LDFLAGS += $(CFLAGS) -fshort-enums -Wl,$(LD_FILE) -Wl,-Map=$@.map -Wl,-cref -Wl,-gc-sections
+
+# Debugging/Optimization
+ifeq ($(DEBUG), 1)
+LDFLAGS += specs=rdimon.specs
```

Comment:
This causes `arm-none-eabi-gcc: error: specs=rdimon.specs: No such file or directory` for me. Sorry but what does this do? never seen this before :). I'm on gcc 9.3.0.